### PR TITLE
Add option for initImages with mount to /init-config

### PR DESCRIPTION
**Pull request by deitch**
**What this PR does / why we need it**:

As discussed in slack with @rfranzke , this PR creates the ability to specify `initImages` in the VPN configuration. This makes it possible to determine one or more configuration components for the VPN dynamically by running an image and saving its output to `/init-config`. As of right now, this is used for Packet, where the node range (and thus the vpn target range) is unknown until the servers are running. In general, it is intended to add any form of dynamic flexibility. 

**Special notes for your reviewer**:

There will be 2 follow-up PRs to this:

1. Documentation, specifying how to use this.
2. vpn image itself, to consume information from a file in `/init-config/`, if it is available.

**Release note**:

```improvement operator
Add support for vpn dynamic configuration via `initContainers`.
```

### CODE REVIEW

**Comment by deitch** [comment]
I will work in those two once this is in.

**Review by rfranzke — CHANGES_REQUESTED** [review]
Thanks! Looks good. Can you please just add an example `initContainers` configuration (commented out) to the [values.yaml](https://github.com/gardener/gardener/blob/master/charts/shoot-core/charts/vpn-shoot/values.yaml) file of this chart?

**Comment by deitch** [comment]
> Can you please just add an example initContainers configuration

Doing it right away.

I realized several things:

1.this structure does not allow any flexibility for a command-line other than the default for the image. We probably should allow an option to specify the command line. Not 100% sure the best way to specify it in that file. 
2. Should we use the `initImages` section I suggested? Or would it be better to make it a subsection of the already existing images?

E.g., current proposal A:

```yml
images:
  vpn-shoot: image-repository:image-tag
initImages:
  - imageA:3.1.2
  - otherImage:1.0.0
serviceNetwork: 10.0.0.0/24
podNetwork: 192.168.0.0/16
nodeNetwork: 172.16.0.0/20
```

Alternate proposal B:

```yml
images:
  vpn-shoot: image-repository:image-tag
  init:
    - imageA:3.1.2
    - otherImage:1.0.0
serviceNetwork: 10.0.0.0/24
podNetwork: 192.168.0.0/16
nodeNetwork: 172.16.0.0/20
```

As for how to specify a command-line, e.g. I want `imageA:3.1.2` but rather than the default, I want to run `ip a eth0 > /init-config/file` or something similar. Why force people to have custom images if a standard one would work with the right command.

It is easy to do in the template, but not sure how I would specify it in the yml. Maybe this?

```yml
images:
  vpn-shoot: image-repository:image-tag
  init:
  - image: imageA:3.1.2
     command: "ip a eth0 > /init-config/file"
  - otherImage:1.0.0
serviceNetwork: 10.0.0.0/24
podNetwork: 192.168.0.0/16
nodeNetwork: 172.16.0.0/20
```

So if the entry is a string, that is the image; if it is a key/value map, we look for `name` for the `image` and `command` keys.

Thoughts on both of these issues?

**Comment by rfranzke** [comment]
Hi @deitch, I would envision sth like this:

```yaml
images:
  vpn-shoot: image-repository:image-tag

serviceNetwork: 10.0.0.0/24
podNetwork: 192.168.0.0/16
nodeNetwork: 172.16.0.0/20

initContainers:
- name: init-{{ $index }}
  image: {{ $image }}
  command:
  - ip a eth0 > /init-config/file
```

WDYT?

**Comment by deitch** [comment]
Sure, that's just as good. Updating this PR now.

**Comment by deitch** [comment]
@rfranzke something like this (now updated commits)?

I did this:

```yml
{{- if .Values.initContainers }}
      initContainers:
      {{- range $index, $ctr := .Values.initContainers }}
      - name: init-{{ $index }}
        image: {{ $ctr.image }}
        {{- if $ctr | .command }}
        command: {{ $ctr.command }}
        {{- end }}
        volumeMounts:
        - name: shared-init-config
          mountPath: /init-config
      {{- end }}
{{- end }}
```

Instead of a straight `{{ toYaml .Values.initContainers }}` since we need to add the following volume mount to each one:

```yml
        volumeMounts:
        - name: shared-init-config
          mountPath: /init-config
```

File: charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml [hunk 1]
```diff
@@ -35,6 +35,19 @@ spec:
       tolerations:
       - key: CriticalAddonsOnly
         operator: Exists
+{{- if .Values.initContainers }}
+      initContainers:
+      {{- range $index, $ctr := .Values.initContainers }}
+      - name: init-{{ $index }}
+        image: {{ $ctr.image }}
+        {{- if $ctr | .command }}
```
Comment (rfranzke) [inline charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml]:
```suggestion
        {{- if $ctr.command }}
```

File: charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml [hunk 2]
```diff
@@ -35,6 +35,19 @@ spec:
...
+        command: {{ $ctr.command }}
```
Comment (rfranzke) [inline charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml]:
```suggestion
        command: {{ $ctr.command }}
        command:
{{ toYaml $ctr.command | indent 8 }}
```

**Review by rfranzke — CHANGES_REQUESTED** [review]
Thanks!

Got error `Error: render error in "vpn-shoot/templates/deployment.yaml": template: vpn-shoot/templates/deployment.yaml:43:22: executing "vpn-shoot/templates/deployment.yaml" at <.command>: command is not a method but has arguments` when trying it out. Made some suggestions to fix this.

File: charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml (hunk 1 above)
Comment (deitch) [reply charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml]: Oh oops. I originally wrote them in with pipes as placeholders, then changed them back, but forgot one. Thanks @rfranzke !

File: charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml (hunk 2 above)
Comment (deitch) [reply charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml]: Would this work if the value of `$ctr.command` is structured as a `["a","b"]` array? I never was sure about it. Will accept as you recommended.

Comment (rfranzke) [reply charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml]: Yes. Try it out: `helm install charts/shoot-core/charts/vpn-shoot --debug --dry-run --name test`

**Comment by deitch** [comment]
I think that addresses it. Any other changes?

File: charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml (hunk 2 above)
Comment (deitch) [reply charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml]: Yup. Thanks.

File: charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml [hunk 3]
```diff
@@ -75,3 +91,5 @@ spec:
       - name: vpn-shoot-dh
         secret:
           secretName: vpn-shoot-dh
+      - name: shared-init-config
+        emptyDir: {}
```
Comment (rfranzke) [inline charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml:95]: I think you should add `emptyDir` to the list of allowed `volumes` for the vpn-shoot pod security policy

**Review by rfranzke — CHANGES_REQUESTED** [review]
One last thing, rest looks good.

File: charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml (hunk 3 above)
Comment (deitch) [reply charts/shoot-core/charts/vpn-shoot/templates/deployment.yaml:95]: Good point, thanks for catching it. Done.

**Comment by deitch** [comment]
Added PSP as suggested.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**